File: list3/python/galois_field.py
def GaloisField(order: int):
  if not (isinstance(order, int)):
    raise ValueError(f"Order={order} is not an integer.")
  if not is_prime(order):
    raise ValueError(f"Order={order} is not prime.")

  class FieldElement():
    # Constructor
    def __init__(self, num: int, /):
      if isinstance(num, FieldElement):
        self.num = num.num
      else:
        self.num = num % order

    # Printing
    def __str__(self):
      return f"GF<{order}>({self.num})"

    # Arithmetic operators
    def __neg__(self):
      return FieldElement(-self.num)

    @typecast
    def __add__(self, other):
      return FieldElement(self.num + other.num)  

    @typecast
    def __sub__(self, other):
      return FieldElement(self.num - other.num)  

    @typecast
    def __mul__(self, other):
      return FieldElement(self.num * other.num)

    @typecast
    def __pow__(self, other):
      return FieldElement(self.num ** other.num)

    @typecast
    def __truediv__(self, other):
      return self * other.mod_inverse()

    @typecast
    def __div__(self, other):
      return self * other.mod_inverse()

    @typecast
    def __divmod__(self, other):
      (quot, remainder) = divmod(self.num, other.num)
      return (FieldElement(quot), FieldElement(remainder))

    def mod_inverse(self):
      return FieldElement(pow(self.num, -1, self.order))
      
    # Comparison operators
    @typecast
    def __eq__(self, other):
      return self.num == other.num

    @typecast
    def __lt__(self, other):
      return self.num < other.num

    @typecast
    def __le__(self, other):
      return self.__lt__(other) or self.__eq__(other)

    @typecast
    def __gt__(self, other):
      return self.num > other.num

    @typecast
    def __ge__(self, other):
      return self.__gt__(other) or self.__eq__(other)

    # Right-side arithmetic operators
    def __radd__(self, other): 
      return self + other

    def __rsub__(self, other): 
      return -self + other
    
    def __rmul__(self, other): 
      return self * other
    
    def __rtruediv__(self, other): 
      return self.mod_inverse() * other
    
    def __rdiv__(self, other): 
      return self.__rtruediv__(other)

  FieldElement.order = order
  return FieldElement

def typecast(f):
  def __cast(self, other):
    if type(self) is not type(other):
      try:
        other = self.__class__(other) # FieldElement(other)
      except Exception as e:
        raise TypeError(
          f"Couldn't cast {other} of type {type(other).__name__} to type " \
          f"{type(self).__name__} in func={f.__name__}.\n" \
          f"Reason: {e}"
        )
    return f(self, other)
  return __cast

def is_prime(num):
  if num == 2 or num == 3: return True
  if num < 2 or num % 2 == 0: return False
  if num < 9: return True
  if num % 3 == 0: return False
  sqrt_num = int(num ** 0.5) # truncated sqrt
  cand = 5
  while cand <= sqrt_num:
    if num % cand == 0: return False
    if num % (cand+2) == 0: return False
    cand += 6
  return True

File: list3/java/GaloisField.java
public class GaloisField implements FieldElement<GaloisField> {
  private long number;
  private long order;

  GaloisField(final long number, final long order) {
    try {
      if (is_prime(order)) {
        this.number = ((((number + order) % order) + order) % order);
        this.order = order;
      } else {
        throw new PrimeException();
      }
      
    } catch (PrimeException e) {
      StackTraceElement frame = new Exception().getStackTrace()[0];
      System.err.println("\n" + frame.getFileName() + ":" + frame.getLineNumber() + ": " + e.getMessage());
      System.exit(0);
    }
  }

  // Arithmetic
  public GaloisField _add(final GaloisField gf) {
    this.number = (((this.number + gf.number + this.order) % this.order) + this.order) % this.order;
    return this;
  }
 
  public GaloisField _add(final long number) {
    this.number = (((this.number + number + this.order) % this.order) + this.order) % this.order;
    return this;
  }

  public GaloisField _sub(final GaloisField gf) {
    this.number = (((this.number - gf.number + this.order) % this.order) + this.order) % this.order;
    return this;
  }
  
  public GaloisField _sub(final long number) {
    this.number = (((this.number - number + this.order) % this.order) + this.order) % this.order;
    return this;
  }

  public GaloisField _mul(final GaloisField gf) {
    this.number = (((this.number * gf.number + this.order) % this.order) + this.order) % this.order;
    return this;
  }
  
  public GaloisField _mul(final long number) {
    this.number = (((this.number * number + this.order) % this.order) + this.order) % this.order;
    return this;
  }

  public GaloisField _div(final GaloisField gf) {
    this._mul(gf._inv());
    return this;
  }

  public GaloisField _div(final long number) {
    this._mul((new GaloisField(number, order))._inv());
    return this;
  }

  // Incrementation
  public void _inc() {
    this.number++;
  }

  public void _dec() {
    this.number--;
  }

  // Comparison
  public boolean _eq(final GaloisField gf) {
    return (this.number == gf.number);
  }

  public boolean _neq(final GaloisField gf) {
    return (this.number != gf.number);
  }

  public boolean _gt(final GaloisField gf) {
    return (this.number > gf.number);
  }

  public boolean _lt(final GaloisField gf) {
    return (this.number < gf.number);
  }

  public boolean _ge(final GaloisField gf) {
    return (this.number >= gf.number);
  }

  public boolean _le(final GaloisField gf) {
    return (this.number <= gf.number);
  }

  // Assignment
  public GaloisField _assign(final long number) {
    this.number = (((number + this.order) % this.order) + this.order) % this.order;
    return this;
  }

  // Equals
  public boolean equals(Object compareObj) {
    if (this == compareObj)
      return true;

    if (!(compareObj instanceof GaloisField))
      return false;

    GaloisField compared = (GaloisField) compareObj;
    if (this.order != compared.order)
      return false;
    if (this.number != compared.number)
      return false;

    return true;
  }

  public GaloisField _neg() {
    return new GaloisField(-this.number, this.order);
  }

  public GaloisField _one() {
    return new GaloisField(1, this.order);
  }

  public GaloisField _zero() {
    return new GaloisField(0, this.order);
  }

  public boolean _isZero() {
    return this.number == 0;
  }

  public boolean _isOne() {
    return this.number == 1;
  }

  // Print
  public String toString() {
    return ("GF<" + this.order + ">(" + this.number + ")");
  }

  // Modular inverse
  private static long mod_inv(final long a, final long b, final long x, final long y) {
    return b == 0 ? x : mod_inv(b, a % b, y, x - y * (a / b));
  }

  public GaloisField _inv() {
    return new GaloisField(mod_inv(this.number, order, 1, 0), order);
  }

  // Getters
  public long _get_number() {
    return number;
  }

  public long _get_order() {
    return order;
  }  

  private static boolean is_prime(final long num) {
    if (num == 2 || num == 3) return true;
    if (num < 2 || num % 2 == 0) return false;
    if (num < 9) return true;
    if (num % 3 == 0) return false;

    long sqrt_num = (long) (Math.sqrt(num));
    long temp = 5;
    while (temp <= sqrt_num) {
      if (num % temp == 0) return false;
      if (num % (temp + 2) == 0) return false;
      temp += 6;
    }
    return true;
  }

  private class PrimeException extends Exception { 
    public PrimeException() {
        super("error: 'order' is not a prime number.");
    }
  }

}
